feat: support Module Federation format (#240)
web-infra-dev/rslib: website/docs/en/guide/advanced/module-federation.mdx

# Module Federation

> This chapter will introduce how to quickly build Module Federation products in Rslib

Reviewer: ```suggestion
This chapter introduces how to build Module Federation output in Rslib.
```

## Glossary

Module Federation is a JavaScript application divide-and-conquer architecture pattern (similar to server-side microservices) that allows you to share code and resources between multiple JavaScript applications (or Micro-Frontend).

## Usage scenarios

Module federation has some typical usage scenarios, including:

> Allows independent applications (called "Micro-Frontend" in the Micro-Frontend architecture) to share modules without having to recompile the entire application.

Reviewer: ```suggestion
- Allows independent applications (called "Micro-Frontend" in the Micro-Frontend architecture) to share modules without having to recompile the entire application.
```

> Different teams work on different parts of the same application without having to recompile the entire application.

Reviewer: ```suggestion
- Different teams work on different parts of the same application without having to recompile the entire application.
```

> Dynamic code loading and sharing between applications at runtime.

Reviewer: ```suggestion
- Dynamic code loading and sharing between applications at runtime.

```

> Module federation can help you:

Reviewer: ```suggestion
Module Federation can help you:
```

- Reduce code duplication
- Improve code maintainability
- Reduce the overall size of the application
- Improve application performance

> ## Quick access

Reviewer: ```suggestion
## Quick Start
```

> First install the Module Federation Rsbuild Plugin (Rslib is built on Rsbuild)

Reviewer: ```suggestion
First install the Module Federation Rsbuild Plugin (Rslib is based on Rsbuild).
```

import { PackageManagerTabs } from '@theme';

> <PackageManagerTabs

Reviewer: Can be simplified like this

```
<PackageManagerTabs command="add @module-federation/rsbuild-plugin -D" />
```

  command={{
    npm: 'npm add @module-federation/rsbuild-plugin --save-dev',
    yarn: 'yarn add @module-federation/rsbuild-plugin --save-dev',
    pnpm: 'pnpm add @module-federation/rsbuild-plugin --save-dev',
    bun: 'bun add @module-federation/rsbuild-plugin --save-dev',
  }}
/>

> Add the plugin to the `rslib.config.ts` file

Reviewer: ```suggestion
Add the plugin to the `rslib.config.ts` file:
```

First install the runtime dependencies

<PackageManagerTabs
  command={{
    npm: 'npm add @module-federation/enhanced --save',
    yarn: 'yarn add @module-federation/enhanced --save',
    pnpm: 'pnpm add @module-federation/enhanced --save',
    bun: 'bun add @module-federation/enhanced --save',
  }}
/>

Then consume other Module Federation modules at runtime, for example

```ts
import { init, loadRemote } from '@module-federation/rsbuild-plugin/runtime';
import { Suspense, createElement, lazy } from 'react';

init({
  name: 'rslib_provider',
  remotes: [
    {
      name: 'mf_remote',
      entry: 'http://localhost:3002/mf-manifest.json',
    },
  ],
});

export const Counter: React.FC = () => {
  return (
    <div>
      <Suspense fallback={<div>loading</div>}>
        {createElement(
          lazy(
            () =>
              loadRemote('mf_remote') as Promise<{
                default: React.FC;
              }>,
          ),
        )}
      </Suspense>
    </div>
  );
};
```

> This ensures that modules can be loaded as expected in multiple formats

Reviewer: ```suggestion
This ensures that modules can be loaded as expected in multiple formats.
```

## FAQs

> If the Rslib producer is built with build, this means that the `process.env.NODE_ENV` of the producer is `production` . If the consumer is started in dev mode at this time,

Reviewer: weird line break

Due to the shared loading policy of Module Federation being `version-first` by default, there may be problems loading into different modes of react and react-dom (e.g. react in development mode, react-dom in production mode).
You can set up [shareStrategy](https://module-federation.io/configure/sharestrategy) at the consumer to solve this problem, but make sure you fully understand this configuration

```ts
pluginModuleFederation({
  // ...
  shareStrategy: 'loaded-first',
}),
```

> ## Reference examples

Reviewer: ```suggestion
## Examples
```
